examples/ldns-notify.1
.TH ldns-notify 1 "9 Jan 2007"
.SH NAME
ldns-notify \- notify DNS servers that updates are available
.SH SYNOPSIS
.B ldns-notify
[options]
\-z zone
.IR servers 

.SH DESCRIPTION

\fBldns-notify\fR sends a NOTIFY message to DNS servers. This tells them
that an updated zone is available at the master servers. It can perform
TSIG signatures and it can add a SOA serial number of the updated zone.
If a server already has that serial number it will disregard the message.

.SH OPTIONS
.TP
\fB-z zone\fR
The zone that is updated.

.TP
\fB-I address\fR
Source IP to send query from.

.TP
\fB-h\fR
Show usage and exit

.TP
\fB-v\fR
Show the version and exit

.TP
\fB-s serial\fR
Append a SOA record indicating the serial number of the updated zone.

.TP
\fB-p port\fR
Use port as destination port (default the DNS port 53) for the UDP packets.

.TP
\fB-y key:data[:algo] \fR
Use the given TSIG key and base64-data, and optinally an algorithm to sign
the NOTIFY. The algorithm defaults to hmac-md5.sig-alg.reg.int.

.TP
\fB-d\fR
Print verbose debug information. The query that is sent and the query
that is received.

.TP
\fB-r num\fR
Specify the maximum number of retries before notify gives up trying to
send the UDP packet.

.SH EXIT CODE
The program exits with a 0 exit code if all servers replied an 
acknowledgement to the notify message, and a failure exit code otherwise. 

.SH AUTHOR
Written by the ldns team as an example for ldns usage.

.SH REPORTING BUGS
Report bugs to <ldns-team@nlnetlabs.nl>. 

.SH COPYRIGHT
Copyright (C) 2005 NLnet Labs. This is free software. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.


Changelog
1.7.1	2019-07-26
	* bugfix: Manage verification paths for OpenSSL >= 1.1.0
	  Thanks Marco Davids
	* bugfix #4106: find the SDK on MacOS X <= 10.6
	  Thanks Bill Cole
	* bugfix #4155: ldns-config contains never used variables
	  Thanks Petr Menšík
	* bugfix #4221: drill -x crashes with malformed IPv4 address 
	  Thanks Oleksandr Tymoshenko
	* bugfix #3437: CDS & CDNSKEY RRsets should be signed with the KSK
	  Thanks Tony Finch
	* bugfix #1566, #1568, #1569, #1570: Potential NULL Dereferences
	  Thanks Bill Parker
	* bugfix #1260: Anticipate strchr returning NULL on unfound char
	  Thanks Stephan Zeisberg
	* bugfix #1257: Free after reallocing to 0 size
	  Thanks Stephan Zeisberg
	* bugfix #1256: Check parse limit before t increment
	  Thanks Stephan Zeisberg
	* bugfix #1245: Only one signature per RRset needs to be valid with
	  ldns-verify-zone.  Thanks Emil Natan.
	* ldns-notify can use all supported hash algorithms with -y.
	* bugfix #1209: make install ldns.pc file
	  Thanks Oleksandr Natalenko
	* bugfix #1218: Only chase DS if signer is parent of owner.
	  Thanks Emil Natan
	* bugfix #617: Retry WKS service and protocol names lower case.
	  Thanks Siali Yan
	* Spelling errors in binaries and man pages
	  Thanks Andreas Schulze
	* removed duplicate condition in ldns_udp_send_query.
	* ldns_wire2pkt: fix null pointer dereference if pkt allocation fails
	  and fix memory leak with more EDNS sections
	  Thanks Jan Vcelak
	* bugfix #1399: ldns_pkt2wire() Python binding is broken.
	  Thanks James Raftery
	* ED25519 and ED448 support. Default is to autodetect support in
	  OpenSSL.  Disable with --disable-ed25519 and --disable-ed448.
	* ldns-notify: can have IPv6 address as argument.
	* Fix time sensitive TSIG compare vulnerability.
	* Fix that ldns-testns ignores sigpipe.
	* Fix that ldns-notify sets the query RR as question RR, this
	  removes the wrong TTL and 0 rdata from the packet printout.
	* Allow -T flag to be used together with drill -x
	* Python bindings compile with swig 4.0
	  Thanks Jitka Plesníková
	* bugfix #4248: drill -DT fails for CNAME domain
	  Thanks Thom Wiggers
	* bugfix #4214: Various fixes and leaks found by coverity.
	  Thanks Petr Menšík 
	* Feature #3394: An -I option to ldns-notify to specify a source
	  IP address to send to notify from.  Thanks Geert Hendrickx
	* Bugfix #279: New API functions ldns_udp_connect2,
	  ldns_tcp_connect2, ldns_udp_bgsend2 and ldns_tcp_bgsend2,
	  that return -1 on failure and allow socket number 0
	  to be returned too.  Thanks Joerg Sonnenberger
	* Bugfix #1447: More verbose reporting of chasing problems with
	  ldns-verify-zone.  Thanks Stephane Guedon
	* OpenSSL engine support with ldns-signzone.
	  See also https://penzin.net/ldns-signzone/
	  Many thanks Vadim Penzin.
	* Various improvements found with shellcheck.
	  Thanks Jeffrey Walton
	* PR #36 Update manpage of ldns-notify to mention algorithm
	  support with TSIG.  Thanks Anand Buddhdev
	* Compile warnings with signed char input to to_lower()
	  and is_digit() with NetBSD.  Thanks Håvard Eidnes
	* Missing Makefile.PL in DNS-LDNS perl module contribution.
	  Thanks Jaap Akkerhuis

1.7.0	2016-12-20
	* Fix lookup of relative names in ldns_resolver_search.
	* bugfix #548: Double free for answers > 4096 in ldns_resolver_send_pkt
	* Follow CNAME's when tracing with drill (TODO dnssec trace)
	* Fix #551 change Regent to Copyright holder in BSD license in
	  some of the headings of the file, to match the opensource.org
	  BSD license.
	* -e option makes ldns-compare-zones exit with status code 2 on difference
	* Filter out specified RR types with ldns-read-zone -e and -E options
	* bugfix #563: Correct DNSKEY from DSA private key. Thanks Peter Koch.
	* bugfix #562: ldns-keygen match DSA key maximum size with library.
	  And check keysizes with all algorithms. Thanks Peter Koch.
	* ldns-verify-zone accepts only one single zonefile as argument.
	* bugfix #573: ldns-keygen write private keys with mode 0600.
	  Thanks Leon Weber
	* Fix configure to make ldns compile with LibreSSL 2.0
	* drill now also accepts dig style -y option
	  (-y <[algo:]name:key> i.s.o. -y <name:key[:algo]>)
	* OPENPGPKEY draft rr types. Enable with: --enable-rrtype-openpgpkey
	* bugfix #608: Correct comment about escaped characters
	* CDS and CDNSKEY rr type from RFC 7344.
	  --enable-rrtype-cds configure option removed
	* fix: Memory leak in ldns_pkt_rr_list_by_name()
	  Thanks Johannes Naab
	* fix: Memory leak in ldns_dname2buffer_wire_compress()
	  Thanks Max Liebkies
	* bugfix #613: Allow tab as whitespace too in last rdata field of types
	  of variable length.  Thanks Xiali Yan
	* bugfix: strip trailing whitespace from $ORIGIN lines in zone files
	* Let ldns-keygen output .ds files only for KSK keys
	* Parse RFC7218 TLSA mnemonics, but do not output them
	* Let ldns-dane use SPKI as the default selector i.s.o. Cert
	* bugfix: Fit left over NSEC3s once more before adding empty non
	  terminals.  Thanks Stuart Browne
	* bugfix #605: Determine default trust anchor location at compile time
	  Thanks Peter Koch
	* bugfix #697: Double free with ldns-dane create
	  Thanks Carsten Strotmann
	* bugfix #623: Do not redefine bool type and boolean values
	  Thanks Jakob Petsovits
	* bugfix #570: Add TLSA, CDS, CDNSKEY and OPENPGPKEY RR types to ldnsx
	  Thanks Shussain
	* bugfix #575: ldns_pkt_clone() does not copy timestamp field
	  Thanks Calle Dybedahl 
	* bugfix #584: ldns-update fixes.  Send update to port 53, bring manpage
	  in sync with the usage text, and don't alter the ldns_resolver passed
	  to ldns_update_soa_zone_mname().  Created a ldns_resolver_clone()
	  function in the process.  Thanks Nicholas Riley.
	* bugfix #633: ldns_pkt_clone() parameter isn't const.
	  Thanks Jakop Petsovits
	* bugfix: ldns-dane manpage correction
	  Thanks Erwin Lansing
	* Spelling fixes.  Thanks Andreas Schulze
	* Hyphen used as minus in manpages.  Thanks Andreas Schulze.
	* RFC7553 RR Type URI is supported by default.

examples/ldns-notify.c
/* 
 * ldns-notify.c - ldns-notify(8)
 * 
 * Copyright (c) 2001-2008, NLnet Labs, All right reserved
 *
 * See LICENSE for the license
 *
 * send a notify packet to a server
 */

#include "config.h"

/* ldns */
#include <ldns/ldns.h>

#ifdef HAVE_NETDB_H
#include <netdb.h>
#endif
#include <errno.h>

static int verbose = 1;
static int max_num_retry = 15; /* times to try */

static void
usage(void)
{
	fprintf(stderr, "usage: ldns-notify [other options] -z zone <servers>\n");
	fprintf(stderr, "Ldns notify utility\n\n");
	fprintf(stderr, " Supported options:\n");
	fprintf(stderr, "\t-z zone\t\tThe zone\n");
	fprintf(stderr, "\t-I <address>\tsource address to query from\n");
	fprintf(stderr, "\t-s version\tSOA version number to include\n");
	fprintf(stderr, "\t-y <name:key[:algo]>\tspecify named base64 tsig key"
	                ", and optional an\n\t\t\t"
	                "algorithm (defaults to hmac-md5.sig-alg.reg.int)\n");
	fprintf(stderr, "\t-p port\t\tport to use to send to\n");
	fprintf(stderr, "\t-v\t\tPrint version information\n");
	fprintf(stderr, "\t-d\t\tPrint verbose debug information\n");
	fprintf(stderr, "\t-r num\t\tmax number of retries (%d)\n", 
		max_num_retry);
	fprintf(stderr, "\t-h\t\tPrint this help information\n\n");
	fprintf(stderr, "Report bugs to <ldns-team@nlnetlabs.nl>\n");
	exit(1);
}

static void
version(void)
{
        fprintf(stderr, "%s version %s\n", PACKAGE_NAME, PACKAGE_VERSION);
        fprintf(stderr, "Written by NLnet Labs.\n\n");
        fprintf(stderr,
                "Copyright (C) 2001-2008 NLnet Labs.  This is free software.\n"
                "There is NO warranty; not even for MERCHANTABILITY or FITNESS\n"
                "FOR A PARTICULAR PURPOSE.\n");
        exit(0);
}

static void
notify_host(int s, struct addrinfo* res, uint8_t* wire, size_t wiresize,
	const char* addrstr)
{
	int timeout_retry = 5; /* seconds */
	int num_retry = max_num_retry;
#ifndef S_SPLINT_S
	fd_set rfds;
#endif
	struct timeval tv;
	int retval = 0;
	ssize_t received = 0;
	int got_ack = 0;
	socklen_t addrlen = 0;
	uint8_t replybuf[2048];
	ldns_status status;
	ldns_pkt* pkt = NULL;
	
	while(!got_ack) {
		/* send it */
		if(sendto(s, (void*)wire, wiresize, 0, 
			res->ai_addr, res->ai_addrlen) == -1) {
			printf("warning: send to %s failed: %s\n",
				addrstr, strerror(errno));
#ifndef USE_WINSOCK
			close(s);
#else
			closesocket(s);
#endif
			return;
		}

		/* wait for ACK packet */
#ifndef S_SPLINT_S
		FD_ZERO(&rfds);
		FD_SET(s, &rfds);
		tv.tv_sec = timeout_retry; /* seconds */
#endif
		tv.tv_usec = 0; /* microseconds */
		retval = select(s + 1, &rfds, NULL, NULL, &tv);
		if (retval == -1) {
			printf("error waiting for reply from %s: %s\n",
				addrstr, strerror(errno));
#ifndef USE_WINSOCK
			close(s);
#else
			closesocket(s);
#endif
			return;
		}
		if(retval == 0) {
			num_retry--;
			if(num_retry == 0) {
				printf("error: failed to send notify to %s.\n",
					addrstr);
				exit(1);
			}
			printf("timeout (%d s) expired, retry notify to %s.\n",
				timeout_retry, addrstr);
		}
		if (retval == 1) {
			got_ack = 1;
		}
	}

	/* got reply */
	addrlen = res->ai_addrlen;
	received = recvfrom(s, (void*)replybuf, sizeof(replybuf), 0,
		res->ai_addr, &addrlen);
	res->ai_addrlen = addrlen;

#ifndef USE_WINSOCK
	close(s);
#else
	closesocket(s);
#endif
	if (received == -1) {
		printf("recv %s failed: %s\n", addrstr, strerror(errno));
		return;
	}

	/* check reply */
	status = ldns_wire2pkt(&pkt, replybuf, (size_t)received);
	if(status != LDNS_STATUS_OK) {
		ssize_t i;
		printf("Could not parse reply packet: %s\n",
			ldns_get_errorstr_by_id(status));
		if (verbose > 1) {
			printf("hexdump of reply: ");
			for(i=0; i<received; i++)
				printf("%02x", (unsigned)replybuf[i]);
			printf("\n");
		}
		exit(1);
	}

	if(verbose) {
		ssize_t i;
		printf("# reply from %s:\n", addrstr);
		ldns_pkt_print(stdout, pkt);
		if (verbose > 1) {
			printf("hexdump of reply: ");
			for(i=0; i<received; i++)
				printf("%02x", (unsigned)replybuf[i]);
			printf("\n");
		}
	}
	ldns_pkt_free(pkt);
}

int
main(int argc, char **argv)
{
	int c;
	int i;

	/* LDNS types */
	ldns_pkt *notify;
	ldns_rr *question;
	ldns_rdf *ldns_zone_name = NULL;
	ldns_status status;
	const char *zone_name = NULL;
	int include_soa = 0;
	uint32_t soa_version = 0;
	int do_hexdump = 1;
	uint8_t *wire = NULL;
	size_t wiresize = 0;
	const char *port = "53";
	char *tsig_sep;
	const char *tsig_name = NULL, *tsig_data = NULL, *tsig_algo = NULL;
	int error;
	struct addrinfo from_hints, *from0 = NULL;

	srandom(time(NULL) ^ getpid());

        while ((c = getopt(argc, argv, "vhdp:r:s:y:z:I:")) != -1) {
                switch (c) {
                case 'd':
			verbose++;
			break;
                case 'p':
			port = optarg;
			break;
                case 'r':
			max_num_retry = atoi(optarg);
			break;
                case 's':
			include_soa = 1;
			soa_version = (uint32_t)atoi(optarg);
			break;
                case 'y':
			if (!(tsig_sep = strchr(optarg, ':'))) {
				printf("TSIG argument is not in form "
					"<name:key[:algo]> %s\n", optarg);
				exit(1);
			}
			tsig_name = optarg;
			*tsig_sep++ = '\0';
			tsig_data = tsig_sep;
			if ((tsig_sep = strchr(tsig_sep, ':'))) {
				*tsig_sep++ = '\0';
				tsig_algo = tsig_sep;
			} else {
				tsig_algo = "hmac-md5.sig-alg.reg.int.";
			}
			/* With dig TSIG keys are also specified with -y,
			 * but format with drill is: -y <name:key[:algo]>
			 *             and with dig: -y [hmac:]name:key
			 *
			 * When we detect an unknown tsig algorithm in algo,
			 * but a known algorithm in name, we cane assume dig
			 * order was used.
			 *
			 * Following if statement is to anticipate and correct
			 * dig order
			 */
			if (strcasecmp(tsig_algo, "hmac-md5.sig-alg.reg.int")&&
			    strcasecmp(tsig_algo, "hmac-md5")                &&
			    strcasecmp(tsig_algo, "hmac-sha1")               &&
			    strcasecmp(tsig_algo, "hmac-sha256")             &&
			    strcasecmp(tsig_algo, "hmac-sha384")             &&
			    strcasecmp(tsig_algo, "hmac-sha512")             &&
			    ! (strcasecmp(tsig_name, "hmac-md5.sig-alg.reg.int")
			    && strcasecmp(tsig_name, "hmac-md5")
			    && strcasecmp(tsig_name, "hmac-sha1")
			    && strcasecmp(tsig_name, "hmac-sha256")
			    && strcasecmp(tsig_name, "hmac-sha384")
			    && strcasecmp(tsig_name, "hmac-sha512"))) {

				/* Roll options */
				const char *tmp_tsig_algo = tsig_name;
				tsig_name = tsig_data;
				tsig_data = tsig_algo;
				tsig_algo = tmp_tsig_algo;
			}	
			printf("Sign with name: %s, data: %s, algorithm: %s\n"
			      , tsig_name, tsig_data, tsig_algo);
			break;
                case 'z':
			zone_name = optarg;
			ldns_zone_name = ldns_dname_new_frm_str(zone_name);
			if(!ldns_zone_name) {
				printf("cannot parse zone name: %s\n", 
					zone_name);
				exit(1);
			}
                        break;
		case 'I':
			memset(&from_hints, 0, sizeof(from_hints));
			from_hints.ai_family = AF_UNSPEC;
			from_hints.ai_socktype = SOCK_DGRAM;
			from_hints.ai_protocol = IPPROTO_UDP;
			from_hints.ai_flags = AI_NUMERICHOST;
			error = getaddrinfo(optarg, 0, &from_hints, &from0);
			if (error) {
				printf("bad address: %s: %s\n", optarg,
					gai_strerror(error));
				exit(EXIT_FAILURE);
			}
			break;
		case 'v':
			version();
			/* fallthrough */
                case 'h':
                case '?':
                default:
                        usage();
                }
        }
        argc -= optind;
        argv += optind;

        if (argc == 0 || zone_name == NULL) {
                usage();
        }

	notify = ldns_pkt_new();
	question = ldns_rr_new();

	if (!notify || !question) {
		/* bail out */
		printf("error: cannot create ldns types\n");
		exit(1);
	}

	/* create the rr for inside the pkt */
	ldns_rr_set_class(question, LDNS_RR_CLASS_IN);
	ldns_rr_set_owner(question, ldns_zone_name);
	ldns_rr_set_type(question, LDNS_RR_TYPE_SOA);
	ldns_rr_set_question(question, true);
	ldns_pkt_set_opcode(notify, LDNS_PACKET_NOTIFY);
	ldns_pkt_push_rr(notify, LDNS_SECTION_QUESTION, question);
	ldns_pkt_set_aa(notify, true);
	ldns_pkt_set_random_id(notify);
	if(include_soa) {
		char buf[10240];
		ldns_rr *soa_rr=NULL;
		ldns_rdf *prev=NULL;
		snprintf(buf, sizeof(buf), "%s 3600 IN SOA . . %u 0 0 0 0",
			zone_name, (unsigned)soa_version);
		/*printf("Adding soa %s\n", buf);*/
		status = ldns_rr_new_frm_str(&soa_rr, buf, 3600, NULL, &prev);
		if(status != LDNS_STATUS_OK) {
			printf("Error adding SOA version: %s\n",
				ldns_get_errorstr_by_id(status));
		}
		ldns_pkt_push_rr(notify, LDNS_SECTION_ANSWER, soa_rr);
	}

	if(tsig_name && tsig_data) {
#ifdef HAVE_SSL
		status = ldns_pkt_tsig_sign(notify, tsig_name,
			tsig_data, 300, tsig_algo, NULL);
		if(status != LDNS_STATUS_OK) {
			printf("Error TSIG sign query: %s\n",
				ldns_get_errorstr_by_id(status));
		}
#else
	fprintf(stderr, "Warning: TSIG needs OpenSSL support, which has not been compiled in, TSIG skipped\n");
#endif
	}

	if(verbose) {
		printf("# Sending packet:\n");
		ldns_pkt_print(stdout, notify);

	}

	status = ldns_pkt2wire(&wire, notify, &wiresize);
	if(wiresize == 0) {
		printf("Error converting notify packet to hex.\n");
		exit(1);
	}

	if(do_hexdump && verbose > 1) {
		printf("Hexdump of notify packet:\n");
		for(i=0; i<(int)wiresize; i++)
			printf("%02x", (unsigned)wire[i]);
		printf("\n");
	}

	for(i=0; i<argc; i++)
	{
		struct addrinfo hints, *res0, *ai_res;
		int default_family = AF_UNSPEC;

		if(verbose)
			printf("# sending to %s\n", argv[i]);
		memset(&hints, 0, sizeof(hints));
		hints.ai_family = default_family;
		/* if(strchr(argv[i], ':')) hints.ai_family = AF_INET6; */
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_protocol = IPPROTO_UDP;
		error = getaddrinfo(argv[i], port, &hints, &res0);
		if (error) {
			printf("skipping bad address: %s: %s\n", argv[i],
				gai_strerror(error));
			continue;
		}
		for (ai_res = res0; ai_res; ai_res = ai_res->ai_next) {
			int s;

			if (from0 && ai_res->ai_family != from0->ai_family)
				continue;

		        s = socket(ai_res->ai_family, ai_res->ai_socktype, 
				ai_res->ai_protocol);
			if(s == -1)
				continue;
			if (from0 && bind(s, from0->ai_addr, from0->ai_addrlen)) {
				perror("Could not bind to source IP");
				exit(EXIT_FAILURE);
			}
			/* send the notify */
			notify_host(s, ai_res, wire, wiresize, argv[i]);
		}
		freeaddrinfo(res0);
	}

	ldns_pkt_free(notify);
	free(wire);
        return 0;
}
